### PR TITLE
(CFACT-199) Implement support for setting fact values from environment.

### PR DESCRIPTION
**Pull request by peterhuene**
Implementing support for setting fact values from environment variables.

Now users can set variables of the form "FACTER_xyz" and a string fact
will be added.  Environment variable facts are the last source of truth
and can override built-in, external, and custom fact values.

This is supported for backwards compatibility with previous versions of
Facter.

### CODE REVIEW

**Comment by MikaelSmith** [comment]
Looks like add_environment_facts test is crashing on Windows. Not immediately clear to me why.

**Comment by peterhuene** [comment]
I see it.  I accidentally passed a modified pointer back to `FreeEnvironmentStringsW`.  Will fix.

**Comment by MikaelSmith** [comment]
Oh yeah, oops.

File: lib/src/ruby/fact.cc [hunk 1]
```diff
@@ -103,15 +105,20 @@ namespace facter { namespace ruby {
             return 0;
         });
 
+        bool add = true;
```
Comment (MikaelSmith) [inline lib/src/ruby/fact.cc]: We should add a test for overriding the RUBY fact specifically.

**Comment by MikaelSmith** [comment]
Extensibility.md needs to be updated for FACTERRUBY.

**Comment by MikaelSmith** [comment]
Also windows/api.cc

File: lib/src/ruby/fact.cc (hunk 1 above)
Comment (peterhuene) [inline lib/src/ruby/fact.cc]: This was just a fix for an unrelated minor bug I noticed where a custom fact overriding a built-in, but returned nil in `setcode` would result in the fact's value changing from "X" to "X" as it just needlessly readds it below.  I can add a test for overriding the ruby fact from the environment if you feel that's necessary, though.

**Comment by peterhuene** [comment]
Whoops, good catch.  I hate how clion only searches through files in the "project" by default, missing text files and files that are excluded based on the platform in CMakelists.txt.  Will fix.

**Comment by peterhuene** [comment]
Updated the `FACTER_RUBY` usages.  Going to do one last push that adds a ruby test for overriding built-ins and overriding custom facts with env vars.

**Comment by peterhuene** [comment]
@MikaelSmith Okay, I've pushed three new ruby tests: test that overriding a built in with a resolver that returns nil doesn't override the built-in, test overriding a built-in (the ruby fact), testing overriding a built-in with a custom fact and then overriding that with an environment variable (last one wins).

File: lib/src/ruby/fact.cc (hunk 1 above)
Comment (peterhuene) [inline lib/src/ruby/fact.cc]: Added a test for that.

**Comment by MikaelSmith** [comment]
Looks good.

File: lib/tests/ruby/ruby.cc [hunk 2]
```diff
@@ -102,13 +103,26 @@ struct ruby_test_parameters
...
+
     string file;
     log_level level;
     bool failure_expected;
     string fact;
     string expected;
+    pair<string, string> variable;
     map<string, string> facts;
     vector<pair<string, string>> messages;
+    vector<pair<string, string>> environment_variables;
```
Comment (peterhuene) [inline lib/tests/ruby/ruby.cc]: This line is dead.  Removing.

**Comment by peterhuene** [comment]
Should be good to merge, I think.

**Pull request merged**